pages/instructorQuestionManualGrading/instructorQuestionManualGrading.js

```diff
@@ -89,58 +72,39 @@ router.get('/', (req, res, next) => {
+     * the 'credit' calculation in a manually graded problem is also problematic, because it means
+     * that the behavior of the instructor editing the score on the manual grading page would be
+     * different than the behavior of the instructor editing the score on any of the other pages
+     * where they can edit score. Fundamentally, we need to rethink how to treat questions that are
+     * manually graded within PrairieLearn and how to handle those score calculations.
+     */
+    sqlDb.call('instance_questions_update_score', params, (err, _result) => {
+      if (ERR(err, next)) return;
+      ltiOutcomes.updateScore(req.body.assessment_instance_id, null, (err) => {
         if (ERR(err, next)) return;
-
-        const { question, variant, submission } = result.rows[0];
-        if (!question || !variant || !submission) {
-          return next(error.make('500', 'Manual grading dependencies missing'));
-        }
-
-        Object.assign(res.locals, { question, variant, submission });
-
-        const params = [
-          req.body.assessment_id,
-          null, // assessment_instance_id,
-          submission.id, // submission_id
-          null, // instance_question_id,
-          null, // uid
-          null, // assessment_instance_number
-          null, // qid
-          score_perc,
-          null, // points
-          { manual: note }, // feedback
-          null, // partial_scores
-          res.locals.authn_user.user_id,
-        ];
-
-        /**
-         *  TODO: calling 'instance_questions_update_score' may not be the perfect thing to do
-         * here, because it won't respect the 'credit' property of the assessment_instance.
-         * However, allowing the 'credit' calculation in a manually graded problem is also problematic,
-         * because it means that the behavior of the instructor editing the score on the manual grading
-         * page would be different than the behavior of the instructor editing the score on any of the other
-         * pages where they can edit score. Fundamentally, we need to rethink how to treat questions
-         * that are manually graded within PrairieLearn and how to handle those score calculations.
-         */
-        sqlDb.call('instance_questions_update_score', params, (err, _result) => {
-          if (ERR(err, next)) return;
-          ltiOutcomes.updateScore(req.body.assessment_instance_id, (err) => {
-            if (ERR(err, next)) return;
-            res.redirect(
-              `${res.locals.urlPrefix}/assessment/${req.body.assessment_id}/assessment_question/${req.body.assessment_question_id}/next_ungraded?instance_question=${res.locals.instance_question.id}`
-            );
-          });
-        });
-      }
-    );
+        res.redirect(
+          `${res.locals.urlPrefix}/assessment/${req.body.assessment_id}/assessment_question/${req.body.assessment_question_id}/next_ungraded?prior_instance_question=${res.locals.instance_question.id}`
```

Comment:
Seems like we could possibly redirect them straight to the next ungraded one instead of requiring a second redirect from the `next_ungraded` route? Not necessary to change now, but may be worth it to add a `TODO:` if you think that's possible.

Reply:
I moved the computation of next ungraded to a lib called by both the next_ungraded route and this redirect. I also async'ed some stuff in the process.